src/renderers/dom/shared/ReactDOMComponent.js

```diff
@@ -992,9 +985,34 @@ ReactDOMComponent.Mixin = {
         DOMPropertyOperations.deleteValueForProperty(getNode(this), propKey);
       }
     }
+
+    var ordered = DOMProperty.order[this._tag];
+    if (ordered != null) {
+      for (var i = 0, len = ordered.length; i < len; i++) {
+        propKey = ordered[i];
+
+        if (nextProps.hasOwnProperty(propKey)) {
+          lastProp = lastProps ? lastProps[propKey] : undefined;
+          nextProp = nextProps[propKey];
+
+          if (lastProp !== nextProp) {
+            DOMPropertyOperations.setValueForProperty(
+              getNode(this),
+              propKey,
+              nextProp
+            );
+          }
+        }
+      }
+    }
+
     for (propKey in nextProps) {
-      var nextProp = nextProps[propKey];
-      var lastProp =
+      if (ordered != null && ordered.indexOf(propKey) >= 0) {
```

Comment:
Formalized my original benchmark and put it on the internet:
https://github.com/nhunzaker/react-ordered-props-bench

I'm happy to accept any profiling tips here. I'm a heavy user of React and benefit from accuracy. I want to get this right.

For your specific question, I added a bunch of properties (custom attributes, if that matters) to some controlled inputs:

```
http://natehunzaker.com/react-ordered-props-bench/heavy-props.html

CONTROL 1628 (+= 2.299877893626147)
EXPERIMENT 1750 (+= 1.870748777520246)

Fastest is EXPERIMENT
```

Though the noise makes it hard to say it's faster, I do not think it is _slower_.

---

Unfortunately, I noticed that there seems to be a performance regression with uncontrolled inputs:

```
http://natehunzaker.com/react-ordered-props-bench/uncontrolled.html

CONTROL 5307 (+= 3.5737538615198474)
EXPERIMENT 4475 (+= 2.2288121746511043)

Fastest is CONTROL
```

Unclear why I didn't catch this before. I'm also not sure specifically why there's a regression. This PR includes fixes to "decontrol" uncontrolled inputs (so that they don't reassign themselves and screw up number inputs). I do not know yet if the performance degradation relates to that, or the ordered properties approach generally. I'm looking in to that now.

---

One other interesting secondary finding is that the `_handleChange` binding always gets applied to uncontrolled inputs, even if there is no change handler:
https://github.com/facebook/react/blob/master/src/renderers/dom/client/wrappers/ReactDOMInput.js#L79

Best I can tell this is waste. Is there a good reason for `_handleChange` to fire on uncontrolled inputs? If not, we could kill some overhead here.